### PR TITLE
Remove all ConstPropNonsense

### PR DESCRIPTION
**Pull request by oli-obk**
We track all locals and projections on them ourselves within the const propagator and only use the InterpCx to actually do some low level operations or read from constants (via `OpTy` we get for said constants).

This helps moving the const prop lint out from the normal pipeline and running it just based on borrowck information. This in turn allows us to make progress on https://github.com/rust-lang/rust/pull/108730#issuecomment-1875557745

there are various follow up cleanups that can be done after this PR (e.g. not matching on Rvalue twice and doing binop checks twice), but lets try landing this one first.

r? @RalfJung

### CODE REVIEW

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 1]
```diff
@@ -258,21 +270,33 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+        }
     }
 
     /// Returns the value, if any, of evaluating `op`. Calls upon `eval_constant`
     /// or `eval_place`, depending on the variant of `Operand` used.
-    fn eval_operand(&mut self, op: &Operand<'tcx>, location: Location) -> Option<OpTy<'tcx>> {
+    fn eval_operand(
+        &mut self,
+        op: &Operand<'tcx>,
+        layout: Option<TyAndLayout<'tcx>>,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:295]: What would be the perf effect of having to recompute the layout?

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 2]
```diff
@@ -515,16 +541,212 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+                ProjectionElem::ConstantIndex { offset, min_length, from_end }
+            }
+            ProjectionElem::Subslice { from, to, from_end } => {
+                ProjectionElem::Subslice { from, to, from_end }
+            }
+            ProjectionElem::Downcast(a, b) => ProjectionElem::Downcast(a, b),
+            ProjectionElem::OpaqueCast(ty) => ProjectionElem::OpaqueCast(ty),
+            ProjectionElem::Subtype(ty) => ProjectionElem::Subtype(ty),
+        })
+    }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:749]: Instead of changing all projections, should `Value::project` take a `&ConstPropagator` to handle the `Index` case, which is quite rare.

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 3]
```diff
@@ -515,16 +541,212 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+    ) -> Option<ProjectionElem<FieldIdx, Ty<'tcx>>> {
+        Some(match proj {
+            ProjectionElem::Deref => ProjectionElem::Deref,
+            ProjectionElem::Field(idx, ty) => ProjectionElem::Field(idx, ty),
+            ProjectionElem::Index(local) => {
+                let val = self.get_const(local.into())?;
+                let op = val.immediate()?;
+                ProjectionElem::Index(FieldIdx::from_u32(
+                    self.ecx.read_target_usize(op).ok()?.try_into().ok()?,
+                ))
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:737]: Should this use `ConstantIndex` instead?

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 4]
```diff
@@ -515,16 +541,212 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+                };
+                let imm = self.use_ecx(|this| {
+                    this.ecx.discriminant_for_variant(
+                        place.ty(this.local_decls(), this.tcx).ty,
+                        variant,
+                    )
+                })?;
+                imm.into()
+            }
+        };
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:717]: We're starting to have quite a few versions of this match on `rvalue`. There are already one in GVN, one in DataflowConstProp. Do you have any idea how we could deduplicate them?

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 5]
```diff
@@ -283,15 +307,16 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+            let val = this.ecx.read_immediate(&arg)?;
             let (_res, overflow) = this.ecx.overflowing_unary_op(op, &val)?;
             Ok((val, overflow))
         })? {
             // `AssertKind` only has an `OverflowNeg` variant, so make sure that is
             // appropriate to use.
             assert_eq!(op, UnOp::Neg, "Neg is the only UnOp that can overflow");
-            let source_info = self.body().source_info(location);
+            let source_info = self.body.source_info(location);
             self.report_assert_as_lint(
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:320]: Drive-by: should `report_assert_as_lint` take a `Location` instead of a `SourceInfo`?

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 6]
```diff
@@ -515,16 +541,212 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
             assert!(
-                self.get_const(local.into()).is_none()
+                matches!(val, Some(Value::Uninit))
                     || self
                         .layout_of(self.local_decls()[local].ty)
                         .map_or(true, |layout| layout.is_zst()),
-                "failed to remove values for `{local:?}`, value={:?}",
-                self.get_const(local.into()),
+                "failed to remove values for `{local:?}`, value={val:?}: {:#?}",
+                self.body,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:551]: Please don't print the body here. For debugging, we should use `-Zdump-mir`.

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:295]: I have not tested this. But I added it for another reason: there are some operations that rely on the layout being passed, and will fail horribly otherwise. I'll remove it so I can actually tell you what these were (and I'll add a comment explaining it then)

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:717]: I have actually copied some logic almost verbatim from GVN and was wondering this myself. It's not clear to me yet if it will be an improvement, but I would like to tackle it so we can at least have an impl that we rejected.

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 6 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:551]: oh whoops, that was just a temporary helper for myself

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:749]: That only works for `project`, but not `project_mut`, so I'd still need this method, and it doesn't appear to have a relevant perf impact

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:734]: I guess I could do that, yea. Trying...
w

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs]: Must have been an intermediate state that resolved itself. Removed the field

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:713]: I don't understand. This method is not used anywhere near `project_mut`, is it?

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs [hunk 7]
```diff
@@ -403,7 +403,12 @@ impl<'a, 'tcx> ConstAnalysis<'a, 'tcx> {
             operand,
             &mut |elem, op| match elem {
                 TrackElem::Field(idx) => self.ecx.project_field(op, idx.as_usize()).ok(),
-                TrackElem::Variant(idx) => self.ecx.project_downcast(op, idx).ok(),
+                TrackElem::Variant(idx) => {
+                    if op.layout.for_variant(&self.ecx, idx).abi.is_uninhabited() {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/dataflow_const_prop.rs:407]: This should be handled when creating the tracked places, skipping recursion in case of uninhabited variant.
Do you have a test-case exhibiting this issue?

Comment (oli-obk) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs:407]: The mir-opt failures from https://github.com/rust-lang/rust/pull/119627#issuecomment-1879047434

File: compiler/rustc_mir_transform/src/const_prop_lint.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs:713]: No, but I wanted to keep the project* methods as similar as possible. I guess they are different enough already for it to not matter

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 8]
```diff
@@ -81,11 +70,85 @@ impl<'tcx> MirLint<'tcx> for ConstPropLint {
...
+                (PlaceElem::Field(idx, _), Value::Aggregate { fields, .. }) => {
+                    fields.get(idx).unwrap_or(&Value::Uninit)
+                }
+                (PlaceElem::Index(idx), Value::Aggregate { fields, .. }) => {
+                    let idx = prop.get_const(idx.into())?.immediate()?;
+                    let idx = prop.ecx.read_target_usize(idx).ok()?;
+                    fields.get(FieldIdx::from_u32(idx.try_into().ok()?)).unwrap_or(&Value::Uninit)
+                }
+                (
+                    PlaceElem::ConstantIndex { offset, min_length: 1, from_end: false },
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:115]: No need to check min_length.

Comment (oli-obk) [reply compiler/rustc_mir_transform/src/const_prop_lint.rs]: we don't actually have any tests exercising `ConstantIndex` in const_prop_lint. I could just remove the entire arm now, too

**Comment by oli-obk** [comment]
r? compiler ralf got enough on his plate and most of this is probably not interesting to him

File: compiler/rustc_mir_transform/src/const_prop.rs
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/const_prop.rs:1]: CanConstProp is only used by const_prop_lint, so IMO it makes more sense to move it over there. Then this file can be removed.

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/const_prop.rs:1]: I was thinking of making this file a support for the other passes that use an interpreter:
- CanConstProp;
- the Machine;
- ...

And eventually common code like arithmetic identities.

**Comment by cjgillot** [comment]
The remaining comments can be addressed in follow-up PRs. I intend to look at the effects on dataflow-const-prop on my side.
@bors r+

File: compiler/rustc_mir_transform/src/const_prop.rs
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/const_prop.rs:1]: That also works, but then it should be called something like `const_prop_utils` and the module-level comment for the file should clarify its purpose.
